### PR TITLE
Cluster membership APIs

### PR DESCRIPTION
**Pull request by mmihic**
Including consistent hashing for resources on to cluster members, with a k8s implementation. Requires that apps run a serviceaccount with a `RoleBinding` permitting `pod [list, watch, get]` within their own namespace.

### CODE REVIEW

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+  ) {
+    enum class Type {
+      // Sent when a cluster changes, contains a watch response
+      CLUSTER_CHANGE,
+
+      // Sent when a new watch is registered, contains a ClusterWatch
+      NEW_WATCH,
+
+      // Sent when a sync point is requested, contains a callback triggered when the event is processed
+      SYNC_POINT,
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt]: for my own knowledge, when is a sync point requested? not following the use case

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    val api = CoreV1Api(client)
+    val connectBackoff = ExponentialBackoff(Duration.ofMillis(100), Duration.ofSeconds(5))
+
+    while (running.get()) {
+      try {
+        log.info { "preparing watch for namespace ${config.my_pod_namespace}" }
+        val watch = Watch.createWatch<V1Pod>(
+            client,
+            api.listNamespacedPodCall(
+                config.my_pod_namespace, // namespace
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt]: lol what an api

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+          clusterService.clusterChanged(response)
+        }
+      } catch (ex: Exception) {
+        // This can occur if we have temporary connectivity glitches to the API server
+        val backoffDelay = connectBackoff.nextRetry()
+        log.error(ex) {
+          "connectivity lost to k8s; waiting ${backoffDelay.toMillis()}ms before retrying"
+        }
+
+        Thread.sleep(backoffDelay.toMillis())
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt]: should successful runs reset the backoff delay?

Comment (mmihic) [reply misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt]: Blerp, yes

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt (hunk 1 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt]: It's a way of saying - call me back when all of the events currently on the queue have been processed. It's mostly for testing. I couldn't think of a good name so if you have a better one I'd be open to it.

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt (hunk 2 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt]: It'd be nice kotlinized with default parameters and such

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt (hunk 1 above)
Comment (ryanhall07) [reply misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt]: ah. well maybe just include that in the comment

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    val connectBackoff = ExponentialBackoff(Duration.ofMillis(100), Duration.ofSeconds(5))
+
+    while (running.get()) {
+      try {
+        log.info { "preparing watch for namespace ${config.my_pod_namespace}" }
+        val watch = Watch.createWatch<V1Pod>(
+            client,
+            api.listNamespacedPodCall(
+                config.my_pod_namespace, // namespace
+                null, // pretty
```
Comment (wesleyk) [inline misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt]: thanks for the comments

File: misk/src/main/kotlin/misk/clustering/ClusterHashRing.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
+package misk.clustering
+
+import com.google.common.hash.HashFunction
+import com.google.common.hash.Hashing
+
+/** A [ClusterHashRing] maps resources to cluster members based on a consistent hash */
+class ClusterHashRing(
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/clustering/ClusterHashRing.kt]: I really like this stuff.

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+      Key.get(Cluster::class.java),
+      Key.get(KubernetesClusterService::class.java)
+  )
+
+  override fun startUp() {
+    log.info { "starting k8s cluster manager" }
+    running.set(true)
+
+    thread(name = "k8s-cluster-event-loop") {
+      runEventLoop()
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt]: I think you wanna use a AbstractExecutionThreadService and override run()

Comment (swankjesse) [reply misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt]: https://github.com/google/guava/wiki/ServiceExplained#abstractexecutionthreadservice

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+  /** Applies an incoming change to a set of cluster members */
+  private fun applyChange(
+    peers: MutableMap<String, Cluster.Member>,
+    changeType: String,
+    pod: V1Pod
+  ): Cluster.Changes {
+    val member = Cluster.Member(pod.metadata.name, pod.status.podIP ?: "")
+
+    // Don't bother to include ourselves in the peer list
+    if (member.name == self.name) return Cluster.Changes(peers.values.toSet())
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt]: neat

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+          Cluster.Changes(peers.values.toSet(), removed = setOf(member))
+        } else Cluster.Changes(peers.values.toSet())
+      }
+      else -> Cluster.Changes(peers.values.toSet())
+    }
+  }
+
+  /**
+   * An internal event handled by the cluster service. All interactions with the cluster service
+   * are dispatched as events polled by the event loop]
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt]: supernit: dangling `]`

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+      CLUSTER_CHANGE,
+
+      // Sent when a new watch is registered, contains a ClusterWatch
+      NEW_WATCH,
+
+      // Sent when a sync point is requested. Allows application code to get a callback when
+      // all of the items currently on the queue have been processed. Useful for testing, to
+      // ensure that we wait until we've reach a well known state where all events we've submitted
+      // have been completed.
+      SYNC_POINT,
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt]: oooh I like it

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt (hunk 2 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt]: I dislike this API ’cause it uses SPDY which is dead tech. Alas.

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            return
+          }
+
+          clusterService.clusterChanged(response)
+        }
+      } catch (ex: Exception) {
+        // This can occur if we have temporary connectivity glitches to the API server
+        val backoffDelay = connectBackoff.nextRetry()
+        log.error(ex) {
+          "connectivity lost to k8s; waiting ${backoffDelay.toMillis()}ms before retrying"
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterWatcher.kt]: nit: just print `$backoffDelay` and let the toString() work out the units?

File: misk/src/test/java/io/kubernetes/client/util/Watches.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package io.kubernetes.client.util;
```
Comment (swankjesse) [inline misk/src/test/java/io/kubernetes/client/util/Watches.java]: package name here wants a comment

File: misk/src/test/kotlin/misk/clustering/kubernetes/KubernetesClusterServiceTest.kt [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+import org.junit.jupiter.api.Test
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+
+internal class KubernetesClusterServiceTest {
+  lateinit var cluster: KubernetesClusterService
+
+  @BeforeEach
+  fun initService() {
+    cluster = KubernetesClusterService(config = KubernetesConfig(
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/clustering/kubernetes/KubernetesClusterServiceTest.kt]: use `@MiskTest` and have the framework handle start + stop?

File: misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt (hunk 6 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt]: oh neat, ok

**Pull request merged**